### PR TITLE
nvidia fp8 support

### PR DESCRIPTION
**Pull request by davidgonmar**
Supports fp8 arithmetic and m16n8k32 tensor cores, with both e4m3 and e5m2 variants.

Arithmetic is supported as a graph rewrite rule that casts fp8 arithmetic to float, and stores the result in fp8 or bool depending on the op.
The approach for tensor cores is similar to existing implementations.

IMPORTANT: FP8 TC support only works on Cuda Toolkit 12.4+ (https://github.com/NVIDIA/cutlass/blob/f93a69134ec8259fd235f220209d6f8734a5cb06/include/cutlass/arch/mma_sm89.h#L53). I tried with a lower version and it did not compile, 12.4 works fine, and CUTLASS also compiles sm_89 tensor core instructions on cuda 12.4+ only.

Any feedback is welcome.

### CODE REVIEW

**Comment by geohot** [comment]
The key thing we need working in fp8 is tensor cores, that's the whole reason we want this, that 600 TFLOPS of perf on a 4090

**Comment by davidgonmar** [comment]
> The key thing we need working in fp8 is tensor cores, that's the whole reason we want this, that 600 TFLOPS of perf on a 4090

yes, once I have the other things fixed, going to work on tensor core support

**Comment by geohot** [comment]
Bounty locked, good luck!

**Comment by davidgonmar** [comment]
> Bounty locked, good luck!

thanks:)

**Comment by davidgonmar** [comment]
Sorry, started the review accidentally.

File: tinygrad/renderer/cstyle.py [hunk 1]
```diff
@@ -347,9 +347,10 @@ def render_kernel(self, function_name, kernel, bufs, uops, prefix=None):
 
     # TODO: this has to be way better to generate for arbitrary M,N,K: use arg[1] for MNK, use arg[4] for vec sizes, encode register packing
     for arg in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
+      ttt = "32" if dt_map[arg[2]][1] == "e4m3" else "16"
       fn, ti, to, ci, co = arg[0], dt_map[arg[2]][0], dt_map[arg[3]][0], dt_map[arg[2]][1], dt_map[arg[3]][1]
       prefix.append(f"""__device__ {to}4 __{fn}({ti}8 a, {ti}4 b, {to}4 c) {{ int *a_pk = (int *) (&a), *b_pk = (int *) (&b);
-asm( "mma.sync.aligned.m16n8k16.row.col.{co}.{ci}.{ci}.{co} {{ %0, %1, %2, %3 }}, {{ %4, %5, %6, %7 }}, {{ %8, %9 }}, {{ %0, %1, %2, %3 }};"
+asm( "mma.sync.aligned.m16n8k{ttt}.row.col.{co}.{ci}.{ci}.{co} {{ %0, %1, %2, %3 }}, {{ %4, %5, %6, %7 }}, {{ %8, %9 }}, {{ %0, %1, %2, %3 }};"
```
[the remark this answers is not in the document]
Comment (flammit) [reply tinygrad/renderer/cstyle.py:353]: should add the layout reference link to `kernel.py` as well: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html?highlight=mma%2520sync%2520aligned#matrix-fragments-for-mma-m16n8k32

**Comment by davidgonmar** [comment]
Got arbitrary fp8 matrices (with correct sizes) to work with tensor cores. Example kernel for (16,32) @ (32,8) matmul:
```c
extern "C" __global__ void __launch_bounds__(32) r_2_2_2_2_2_8_2_2_2_2(float* data0, const __nv_fp8_e4m3* data1, const __nv_fp8_e4m3* data2) {                                          
  int lidx0 = threadIdx.x; /* 8 */                                                                                                                                                      
  int lidx1 = threadIdx.y; /* 2 */                                                                                                                                                      
  int lidx2 = threadIdx.z; /* 2 */                                                                                                                                                      
  int alu0 = (lidx0/4);                                                                                                                                                                 
  int alu1 = (lidx0%2);                                                                                                                                                                 
  int alu2 = ((lidx0/2)%2);                                                                                                                                                             
  int alu3 = ((alu1*2)+(alu2*4)+(alu0*16)+(lidx1*32)+(lidx2*64));                                                                                                                       
  int alu4 = ((alu1*4)+(alu2*8)+(alu0*64)+(lidx1*128)+(lidx2*256));                                                                                                                     
  int alu5 = ((alu1*32)+(alu2*64)+alu0+(lidx1*2)+(lidx2*4));                                                                                                                            
  __nv_fp8_e4m3 val0 = data1[alu4+1];                                                                                                                                                   
  __nv_fp8_e4m3 val1 = data1[alu4+2];                                                                                                                                                   
  __nv_fp8_e4m3 val2 = data1[alu4+3];                                                                                                                                                   
  __nv_fp8_e4m3 val3 = data1[alu4+16];                                                                                                                                                  
  __nv_fp8_e4m3 val4 = data1[alu4+17];                                                                                                                                                  
  __nv_fp8_e4m3 val5 = data1[alu4+18];                                                                                                                                                  
  __nv_fp8_e4m3 val6 = data1[alu4+19];                                                                                                                                                  
  __nv_fp8_e4m3 val7 = data1[alu4+32];                                                                                                                                                  
  __nv_fp8_e4m3 val8 = data1[alu4+33];                                                                                                                                                  
  __nv_fp8_e4m3 val9 = data1[alu4+34];                                                                                                                                                  
  __nv_fp8_e4m3 val10 = data1[alu4+35];                                                                                                                                                 
  __nv_fp8_e4m3 val11 = data1[alu4+48];                                                                                                                                                 
  __nv_fp8_e4m3 val12 = data1[alu4+49];                                                                                                                                                 
  __nv_fp8_e4m3 val13 = data1[alu4+50];                                                                                                                                                 
  __nv_fp8_e4m3 val14 = data1[alu4+51];                                                                                                                                                 
  __nv_fp8_e4m3 val15 = data1[alu4];
  __nv_fp8_e4m3 val16 = data2[alu5+8];
  __nv_fp8_e4m3 val17 = data2[alu5+16];
  __nv_fp8_e4m3 val18 = data2[alu5+24];
  __nv_fp8_e4m3 val19 = data2[alu5+128];
  __nv_fp8_e4m3 val20 = data2[alu5+136];
  __nv_fp8_e4m3 val21 = data2[alu5+144];
  __nv_fp8_e4m3 val22 = data2[alu5+152];
  __nv_fp8_e4m3 val23 = data2[alu5];
  float4 wmma0 = __WMMA_8_16_32_f8e4m3_float(make___nv_fp8_e4m316(val15,val0,val1,val2,val7,val8,val9,val10,val3,val4,val5,val6,val11,val12,val13,val14), make___nv_fp8_e4m38(val23,val1
6,val17,val18,val19,val20,val21,val22), make_float4(0.0f,0.0f,0.0f,0.0f));
  *((float2*)(data0+alu3+8)) = make_float2(wmma0.z,wmma0.w);
  *((float2*)(data0+alu3)) = make_float2(wmma0.x,wmma0.y);
}
```
Saw around 67TFLOPS on a big matrix, will need to test more sizes

**Comment by davidgonmar** [comment]
I assume the bounty title means also PTX support. I think it's better to merge this first and I'll submit another PR with PTX support (should be trivial).

File: tinygrad/runtime/support/compiler_cuda.py [hunk 2]
```diff
@@ -46,10 +46,12 @@ def nv_disassemble(lib):
   except Exception as e: print("Failed to disasm cubin:", str(e), "Make sure your PATH contains nvdisasm binary of compatible version.")
 
 class CUDACompiler(Compiler):
+  version: Tuple[int, int]
```
Comment (nimlgen) [inline tinygrad/runtime/support/compiler_cuda.py:49]: just `self.version` should be good, no?

File: tinygrad/runtime/support/compiler_cuda.py [hunk 3]
```diff
@@ -49,7 +49,8 @@ class CUDACompiler(Compiler):
   def __init__(self, arch:str, cache_key:str="cuda"):
     self.arch, self.compile_options = arch, [f'--gpu-architecture={arch}', "-I/usr/local/cuda/include", "-I/usr/include", "-I/opt/cuda/include/"]
     nvrtc_check(nvrtc.nvrtcVersion((nvrtcMajor := ctypes.c_int()), (nvrtcMinor := ctypes.c_int())))
-    if (nvrtcMajor.value, nvrtcMinor.value) >= (12, 4): self.compile_options.append("--minimal")
+    self.version = (nvrtcMajor.value, nvrtcMinor.value)
```
Comment (geohot) [inline tinygrad/runtime/support/compiler_cuda.py:52]: If you need functionality like this, it should be in its own PR

File: test/test_dtype.py [hunk 4]
```diff
@@ -31,11 +31,12 @@ def _test_to_np(a:Tensor, np_dtype, target):
   except AssertionError as e:
     raise AssertionError(f"\ntensor {a.numpy()} does not match target {target} with np_dtype {np_dtype}") from e
 
-def _assert_eq(tensor:Tensor, target_dtype:DType, target):
+def _assert_eq(tensor:Tensor, target_dtype:DType, target, rtol=None):
```
Comment (geohot) [inline test/test_dtype.py:34]: This too, why is this in here it's a new feature

**Comment by geohot** [comment]
This PR contains several things that should be standalone PRs. To be a standalone PR, it must provide some standalone functionality!

**Comment by davidgonmar** [comment]
> This PR contains several things that should be standalone PRs. To be a standalone PR, it must provide some standalone functionality!

ok
